### PR TITLE
client-side support for accepting TOU

### PR DESCRIPTION
**Pull request by singular-value**
Adding Palash & Rich as same reviewers from server-side

### CODE REVIEW

File: general_superstaq/user_config.py [hunk 1]
```diff
@@ -22,6 +22,10 @@ def get_balance(self, pretty_output: bool = True) -> Union[str, float]:
             return f"${balance:,.2f}"
         return balance
 
+    def accept_terms_of_use(self) -> str:
+        """Confirm acceptance of terms of use at https://superstaq.super.tech/terms_of_use."""
+        return self._client.accept_terms_of_use()
```
Comment (richrines1) [inline general_superstaq/user_config.py:27]: should we make the user type something in a y/n prompt/popup/something to confirm? it seems a bit sketchy that just calling the function automatically indicates acceptance (personally i can imagine running it just to see what it does/expecting a popup or to be taken to a webpage, and then being surprised to find that that means i already accepted)

Comment (singular-value) [reply general_superstaq/user_config.py]: As discussed, good idea.

This now works—test that I ran below, just removed the local server warnings.

```
In [1]: import qiskit_superstaq as qss

In [2]: provider = qss.SuperstaQProvider(remote_host="https://127.0.0.1:5000")

([3] was irrelevant)
 
In [4]: provider.get_balance()
Acceptance of the Terms of Use (superstaq.super.tech/terms_of_use) is necessary before using SuperstaQ.
Type in YES to accept: no
---------------------------------------------------------------------------
SuperstaQException                        Traceback (most recent call last)
Cell In[4], line 1
----> 1 provider.get_balance()

File ~/Developer/general-superstaq/general_superstaq/user_config.py:20, in UserConfig.get_balance(self, pretty_output)
     11 def get_balance(self, pretty_output: bool = True) -> Union[str, float]:
     12     """Get the querying user's account balance in USD.
     13     Args:
     14         pretty_output: whether to return a pretty string or a float of the balance.
   (...)
     18             returns a float of the balance.
     19     """
---> 20     balance = self._client.get_balance()["balance"]
     21     if pretty_output:
     22         return f"${balance:,.2f}"

File ~/Developer/general-superstaq/general_superstaq/superstaq_client.py:177, in _SuperstaQClient.get_balance(self)
    171 def get_balance(self) -> Dict[str, float]:
    172     """Get the querying user's account balance in USD.
    173 
    174     Returns:
    175         The json body of the response as a dict.
    176     """
--> 177     return self.get_request("/balance")

File ~/Developer/general-superstaq/general_superstaq/superstaq_client.py:103, in _SuperstaQClient.get_request(self, endpoint)
     96 def request() -> requests.Response:
     97     return requests.get(
     98         f"{self.url}{endpoint}",
     99         headers=self.headers,
    100         verify=self.verify_https,
    101     )
--> 103 return self._make_request(request).json()

File ~/Developer/general-superstaq/general_superstaq/superstaq_client.py:342, in _SuperstaQClient._make_request(self, request)
    339     if response.ok:
    340         return response
--> 342     self._handle_status_codes(response)
    343     message = response.reason
    345 # Fallthrough should retry.

File ~/Developer/general-superstaq/general_superstaq/superstaq_client.py:287, in _SuperstaQClient._handle_status_codes(self, response)
    285 if response.status_code == requests.codes.unauthorized:
    286     if "terms of use" in str(response.text).lower():
--> 287         self._prompt_accept_terms_of_use()
    288         return
    289     else:

File ~/Developer/general-superstaq/general_superstaq/superstaq_client.py:316, in _SuperstaQClient._prompt_accept_terms_of_use(self)
    314     print("Accepted. You can now continue using SuperstaQ.")
    315 else:
--> 316     raise gss.SuperstaQException(
    317         "You'll need to accept Terms of Use before usage of SuperstaQ.",
    318         requests.codes.unauthorized,
    319     )

SuperstaQException: Status code: 401, Message: 'You'll need to accept Terms of Use before usage of SuperstaQ.'

In [5]: provider.get_balance()
Acceptance of the Terms of Use (superstaq.super.tech/terms_of_use) is necessary before using SuperstaQ.
Type in YES to accept: YES
Accepted. You can now continue using SuperstaQ.
Out[5]: '$208.62'
```

File: general_superstaq/superstaq_client.py [hunk 2]
```diff
@@ -285,11 +289,15 @@ def aqt_get_configs(self) -> Dict[str, str]:
 
     def _handle_status_codes(self, response: requests.Response) -> None:
         if response.status_code == requests.codes.unauthorized:
-            raise gss.SuperstaQException(
-                '"Not authorized" returned by SuperstaQ API.  '
-                "Check to ensure you have supplied the correct API key.",
-                response.status_code,
-            )
+            if "terms of use" in str(response.text).lower():
```
Comment (richrines1) [inline general_superstaq/superstaq_client.py:299]: why can't we compare the response text exactly?

File: general_superstaq/superstaq_client.py [hunk 3]
```diff
@@ -301,6 +316,21 @@ def _handle_status_codes(self, response: requests.Response) -> None:
                 response.status_code,
             )
 
+    def _prompt_accept_terms_of_use(self) -> None:
+        message = (
+            "Acceptance of the Terms of Use (superstaq.super.tech/terms_of_use)"
+            " is necessary before using SuperstaQ.\nType in YES to accept: "
+        )
+        user_input = input(message).upper()
+        if user_input == "YES":
```
Comment (richrines1) [inline general_superstaq/superstaq_client.py:325]: imo this should be checked server side (by sending the user's response text to the server and waiting for a positive response) to make it unambiguous that the acceptance was intentional

File: general_superstaq/user_config_test.py [hunk 4]
```diff
@@ -21,6 +21,18 @@ def test_service_get_balance() -> None:  # pylint: disable=missing-function-docs
     assert service.get_balance(pretty_output=False) == 12345.6789
 
 
+def test_accept_terms_of_use() -> None:  # pylint: disable=missing-function-docstring
```
Comment (singular-value) [inline general_superstaq/user_config_test.py:24]: Btw @pgoiporia what's the lowdown on this?
Could all tests be excluded from needing docstring? Feel free to point me elsewhere if there was a discussion on the PR.

File: general_superstaq/superstaq_client.py (hunk 3 above)
Comment (singular-value) [reply general_superstaq/superstaq_client.py:321]: Good point—updated.

File: general_superstaq/superstaq_client.py (hunk 2 above)
Comment (singular-value) [reply general_superstaq/superstaq_client.py:295]: Done in next version.

File: general_superstaq/superstaq_client.py [hunk 5]
```diff
@@ -281,11 +292,17 @@ def aqt_get_configs(self) -> Dict[str, str]:
 
     def _handle_status_codes(self, response: requests.Response) -> None:
         if response.status_code == requests.codes.unauthorized:
-            raise gss.SuperstaQException(
-                '"Not authorized" returned by SuperstaQ API.  '
-                "Check to ensure you have supplied the correct API key.",
-                response.status_code,
-            )
+            if str(response.text) == (
```
Comment (richrines1) [inline general_superstaq/superstaq_client.py:295]: does this work?
```suggestion
            if response.text == (
```
or does it cause a mypy issue

File: general_superstaq/superstaq_client.py [hunk 6]
```diff
@@ -297,6 +314,20 @@ def _handle_status_codes(self, response: requests.Response) -> None:
...
             )
 
+    def _prompt_accept_terms_of_use(self) -> None:
+        message = (
+            "Acceptance of the Terms of Use (superstaq.super.tech/terms_of_use)"
+            " is necessary before using SuperstaQ.\nType in YES to accept: "
+        )
+        user_input = input(message).upper()
+        response = self._accept_terms_of_use(user_input)
+        print(response)
```
Comment (richrines1) [inline general_superstaq/superstaq_client.py:324]: definitely sufficient for now, but we might want to look into making this a popup when not in a python shell

File: general_superstaq/user_config_test.py [hunk 7]
```diff
@@ -21,6 +21,18 @@ def test_service_get_balance() -> None:  # pylint: disable=missing-function-docs
...
+    client = gss.superstaq_client._SuperstaQClient(
+        remote_host="http://example.com", api_key="key", client_name="general_superstaq"
+    )
+    service = gss.user_config.UserConfig(client)
+    with mock.patch(
+        "general_superstaq.superstaq_client._SuperstaQClient.post_request"
+    ) as mock_post_request:
+        service._accept_terms_of_use("YES")
+        mock_post_request.assert_called_once_with("/accept_terms_of_use", {"user_input": "YES"})
```
Comment (richrines1) [inline general_superstaq/user_config_test.py:33]: nit
```suggestion
        service._accept_terms_of_use("response")
        mock_post_request.assert_called_once_with("/accept_terms_of_use", {"user_input": "response"})
```

File: general_superstaq/superstaq_client.py [hunk 8]
```diff
@@ -178,6 +178,17 @@ def get_balance(self) -> Dict[str, float]:
         """
         return self.get_request("/balance")
 
+    def _accept_terms_of_use(self, user_input: str) -> str:
+        """Makes a POST request to SuperstaQ API to confirm acceptance of terms of use.
+
+        Returns:
+            A tuple.
+
```
Comment (richrines1) [inline general_superstaq/superstaq_client.py:186]: should be replaced with "Args:" description

File: general_superstaq/superstaq_client_test.py [hunk 9]
```diff
@@ -108,9 +108,44 @@ def test_superstaq_client_attributes() -> None:
...
+    mock_accept_terms_of_use.return_value = "YES response required to proceed"
+
+    with mock.patch("builtins.input", return_value="NO"):
+        with pytest.raises(gss.SuperstaQException, match="You'll need to accept Terms of Use"):
+            client.get_balance()
+        assert capsys.readouterr().out == "YES response required to proceed\n"
+
+    # When user accepts, a second request will be made (which would presumably succeed). Strange
+    # mock syntax, see docs.python.org/3.11/library/unittest.mock.html#unittest.mock.PropertyMock
+    type(fake_get_response).ok = mock.PropertyMock(side_effect=[False, True])
```
Comment (richrines1) [inline general_superstaq/superstaq_client_test.py:139]: this should also work an imo is easier to follow:
```suggestion
    fake_authorized_get_response = mock.MagicMock(ok=True)
    mock_get.side_effect = [fake_get_response, fake_authorized_get_response]
```

File: general_superstaq/superstaq_client_test.py [hunk 10]
```diff
@@ -108,9 +108,44 @@ def test_superstaq_client_attributes() -> None:
...
+
+    fake_get_response = mock.MagicMock()
+    fake_get_response.ok = False
+    fake_get_response.status_code = requests.codes.unauthorized
+    fake_get_response.text = "You must accept the Terms of Use (superstaq.super.tech/terms_of_use)."
+    mock_get.return_value = fake_get_response
+
+    mock_accept_terms_of_use.return_value = "YES response required to proceed"
+
+    with mock.patch("builtins.input", return_value="NO"):
```
Comment (richrines1) [inline general_superstaq/superstaq_client_test.py:132]:
```suggestion
    with mock.patch("builtins.input"):
```

File: general_superstaq/superstaq_client_test.py [hunk 11]
```diff
@@ -108,9 +108,44 @@ def test_superstaq_client_attributes() -> None:
...
+        with pytest.raises(gss.SuperstaQException, match="You'll need to accept Terms of Use"):
+            client.get_balance()
+        assert capsys.readouterr().out == "YES response required to proceed\n"
+
+    # When user accepts, a second request will be made (which would presumably succeed). Strange
+    # mock syntax, see docs.python.org/3.11/library/unittest.mock.html#unittest.mock.PropertyMock
+    type(fake_get_response).ok = mock.PropertyMock(side_effect=[False, True])
+    mock_accept_terms_of_use.return_value = "Accepted. You can now continue using SuperstaQ."
+    with mock.patch("builtins.input", return_value="YES"):
```
Comment (richrines1) [inline general_superstaq/superstaq_client_test.py:141]:
```suggestion
    with mock.patch("builtins.input"):
```

File: general_superstaq/user_config_test.py (hunk 4 above)
Comment (vtomole) [reply general_superstaq/user_config_test.py:24]: Related: https://github.com/SupertechLabs/general-superstaq/pull/234#issuecomment-1442384088

File: general_superstaq/superstaq_client.py (hunk 5 above)
Comment (singular-value) [reply general_superstaq/superstaq_client.py:295]: Cool yeah that works, thanks.

File: general_superstaq/superstaq_client.py (hunk 8 above)
Comment (singular-value) [reply general_superstaq/superstaq_client.py]: Done

**Review by richrines1 — APPROVED** [review]
lgtm!

**Pull request closed without merging**